service/history/shard/task_key_generator.go

```diff
@@ -0,0 +1,207 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package shard
+
+import (
+	"fmt"
+	"time"
+
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/util"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+const (
+	taskIDUninitialized = -1
+)
+
+type (
+	renewRangeIDFn func() error
+
+	taskKeyGenerator struct {
+		nextTaskID         int64
+		exclusiveMaxTaskID int64
+
+		taskMinScheduledTime time.Time
+
+		rangeSizeBits uint
+		timeSource    clock.TimeSource
+		logger        log.Logger
+
+		renewRangeIDFn renewRangeIDFn
```

Comment:
So the renewRangeIDFn needs to invoke setRangeID()? Could it return the range id and we control the setRangeID()?

Reply:
yes. renewRangeIDFn needs to call setRangeID, like what I did in the test.

I was also think about returning new rangeID from the func and let generator itself handle the rangeID update yesterday. However, I realized that that won't eliminate the need of setRangeID() method on generator because shard can decide to use a new rangeID at any time. 

It's possible to write the code in a way that for this case of renew (exhausted taskID case), let generator handle the rangeID update, while in other cases of renew (e.g initial shard acquire, handle write errors), let the renew rangeID logic in shardContext to set the rangeID. But that means in the renewRangeID logic in shardContext we need to have a way of knowing who triggers the renew and based on that determine if setRangeID() on the generator should be called. So I figured it's probably better to have only one path for updating rangeID: from outside the generator, instead of both within & outside generator. 

But let me know if you have a better idea.

Reply:
is there any issue if the shard context update the rangeID but the task id generator still uses the previous range id?

Reply:
Synced offline. The logic in this PR and follow up changes in shard context (link in PR desc), will make sure generated taskIDs always belong to the current shard rangeID. Same as the logic today. Does that have to be the case? probably no, but I don't want to make that part of the PR and "No shard lock on I/O" work, the risk is too high.